### PR TITLE
add conditional set-content functions

### PR DESCRIPTION
**Pull request by matthewdunsdon**
:thought_balloon:  Often when writing d3 components; it is helpful to think about conditionally setting content, either when component is known to be given an empty DOM node or when component needs to replace to content of a DOM node.

## Description

Introduced functions `setContentIfEmpty(content)` and `setContentIfMissing(content, selector)` to conditionally set content, either based upon whether the selector has no children nodes or whether a supplied selector matches any DOM nodes.

These functions have been designed to be like `append-if-missing`, it that:

- when called from a `selection.select(...)`, they _conditionally_ modify the DOM and return a new selection.
- when called from a `selection.each(...)`, a simular effect is achieved but the new selection is ignored.

## Motivation and Context

Please take a look at <https://github.com/zambezi/d3-utils/issues/14> for more details on this.

## How Was This Tested?
Manually using a "npm run build" and opening `file:///C:/_DEV_/d3-utils/examples/browser.html` in Chrome 56.0.2924.76.

If we want to check in more browsers, it would be easier if we could get a pre-release build published, and I could try it with more browsers.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My change follows the style of this project
- [x] My change requires a change to the documentation
- [x] I have updated the documentation accordingly
- [x] I have read the [contribution guidelines](../CONTRIBUTING.md)

### CODE REVIEW

**Comment by matthewdunsdon** [comment]
@gabrielmontagne, @Poetro; It would be great to get some feedback regarding this PR - _especially the man pages as I am not sure if the explanations are clear enough._ :thinking:

File: man/set-content-if-empty.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
+## set content if empty
+
+Creates a function that when passed to d3's `selection.select` operator will conditionally set the HTML content of a node (only if it is empty) and select the first child DOM element it finds.
```
Comment (mstade) [inline man/set-content-if-empty.md:3]: The "first child element" note has me confused, does this mean the template doesn't necessarily have to have a single root node? If the requirement is that the template has a single root node, then we can rephrase this as such:

> Creates a function that when passed to d3's `selection.select` operator will set the HTML content of a node *only if* it is *empty* and select the root node of the supplied template.

File: man/set-content-if-missing.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
+## set content if missing
+
+Creates a function that when passed to d3's `selection.select` operator will conditionally set the HTML content of a node (only if it doesn't contain something that matches the provided selector) and makes a selection using the selector.
```
Comment (mstade) [inline man/set-content-if-missing.md:3]: Like above, if the template must contain only a single root node, we can rephrase this to be a bit more descriptive I think:

> Creates a function that when passed to d3's `selection.select` operator will set the contents of a node *only if* it *does not* match the provided selector, and selects the root node of the supplied template.

**Review by gabrielmontagne — APPROVED** [review]
Hey @matthewdunsdon , sorry for the delay.
This looks great : )

**Comment by gabrielmontagne** [comment]
I can bake a release as soon as we're all happy.

File: man/set-content-if-empty.md (hunk 1 above)
Comment (matthewdunsdon) [reply man/set-content-if-empty.md:3]: Yes, the current implementation allows you to use templates with multiple top level nodes.

```javascript
 const template = `
<header class="headings">
  <h2 class="greeting">hello</h2>
</header>
<div class="frame">
  <div class="content" />
</div>
`

target
  .select(setContentIfEmpty(template))
    .style('background-color', 'plum')

target
  .select('.content')
     .text(textContent)

 ```
 
 I think we should allow the user to do this, even if it is not the common case.
